File: examples/python/tree_dfs_examples.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# 1. Preorder Traversal
def preorder_traversal(root):
    result = []
    def dfs(node):
        if node:
            result.append(node.val)
            dfs(node.left)
            dfs(node.right)
    dfs(root)
    return result

# 2. Inorder Traversal
def inorder_traversal(root):
    result = []
    def dfs(node):
        if node:
            dfs(node.left)
            result.append(node.val)
            dfs(node.right)
    dfs(root)
    return result

# 3. Postorder Traversal
def postorder_traversal(root):
    result = []
    def dfs(node):
        if node:
            dfs(node.left)
            dfs(node.right)
            result.append(node.val)
    dfs(root)
    return result

# 4. Max Depth
def max_depth(root):
    if not root:
        return 0
    return 1 + max(max_depth(root.left), max_depth(root.right))

# 5. Is Same Tree
def is_same_tree(p, q):
    if not p and not q:
        return True
    if not p or not q or p.val != q.val:
        return False
    return is_same_tree(p.left, q.left) and is_same_tree(p.right, q.right)

# 6. Is Symmetric
def is_symmetric(root):
    def is_mirror(left, right):
        if not left and not right:
            return True
        if not left or not right or left.val != right.val:
            return False
        return is_mirror(left.left, right.right) and is_mirror(left.right, right.left)
    return not root or is_mirror(root.left, root.right)

# 7. Has Path Sum
def has_path_sum(root, target_sum):
    if not root:
        return False
    if not root.left and not root.right:
        return root.val == target_sum
    return (has_path_sum(root.left, target_sum - root.val) or 
            has_path_sum(root.right, target_sum - root.val))

# 8. Diameter Of Binary Tree
def diameter_of_binary_tree(root):
    max_diameter = 0
    def dfs(node):
        nonlocal max_diameter
        if not node:
            return 0
        left = dfs(node.left)
        right = dfs(node.right)
        max_diameter = max(max_diameter, left + right)
        return 1 + max(left, right)
    dfs(root)
    return max_diameter

# 9. Lowest Common Ancestor
def lowest_common_ancestor(root, p, q):
    if not root or root == p or root == q:
        return root
    left = lowest_common_ancestor(root.left, p, q)
    right = lowest_common_ancestor(root.right, p, q)
    return root if left and right else left or right

# 10. Binary Tree Paths
def binary_tree_paths(root):
    result = []
    def dfs(node, path):
        if node:
            path += str(node.val)
            if not node.left and not node.right:
                result.append(path)
            else:
                path += "->"
                dfs(node.left, path)
                dfs(node.right, path)
    dfs(root, "")
    return result

# 11. Is Valid Bst
def is_valid_bst(root):
    def validate(node, min_val, max_val):
        if not node:
            return True
        if node.val <= min_val or node.val >= max_val:
            return False
        return (validate(node.left, min_val, node.val) and 
                validate(node.right, node.val, max_val))
    return validate(root, float('-inf'), float('inf'))

# 12. Sum Root To Leaf Numbers
def sum_root_to_leaf_numbers(root):
    def dfs(node, current_sum):
        if not node:
            return 0
        current_sum = current_sum * 10 + node.val
        if not node.left and not node.right:
            return current_sum
        return dfs(node.left, current_sum) + dfs(node.right, current_sum)
    return dfs(root, 0)

# 13. Run Examples
def run_examples():
    print("=== Tree DFS Examples ===")
    
    # Create test tree: [3,9,20,null,null,15,7]
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    
    # Test 1: Preorder Traversal
    print(f"1. Preorder: {preorder_traversal(root)}")
    
    # Test 2: Inorder Traversal
    print(f"2. Inorder: {inorder_traversal(root)}")
    
    # Test 3: Postorder Traversal
    print(f"3. Postorder: {postorder_traversal(root)}")
    
    # Test 4: Max Depth
    print(f"4. Max Depth: {max_depth(root)}")
    
    # Test 5: Same Tree
    root2 = TreeNode(3)
    root2.left = TreeNode(9)
    root2.right = TreeNode(20)
    print(f"5. Is Same Tree: {is_same_tree(root, root2)}")
    
    # Test 6: Symmetric Tree
    sym_root = TreeNode(1)
    sym_root.left = TreeNode(2)
    sym_root.right = TreeNode(2)
    print(f"6. Is Symmetric: {is_symmetric(sym_root)}")
    
    # Test 7: Path Sum
    print(f"7. Has Path Sum (12): {has_path_sum(root, 12)}")
    
    # Test 8: Diameter
    print(f"8. Diameter: {diameter_of_binary_tree(root)}")
    
    # Test 9: LCA
    p = root.left
    q = root.right.left
    lca = lowest_common_ancestor(root, p, q)
    print(f"9. LCA of 9 and 15: {lca.val if lca else None}")
    
    # Test 10: Binary Tree Paths
    print(f"10. Binary Tree Paths: {binary_tree_paths(root)}")
    
    # Bonus: Valid BST
    bst_root = TreeNode(2)
    bst_root.left = TreeNode(1)
    bst_root.right = TreeNode(3)
    print(f"Valid BST: {is_valid_bst(bst_root)}")
    
    # Bonus: Sum Root to Leaf
    num_root = TreeNode(1)
    num_root.left = TreeNode(2)
    num_root.right = TreeNode(3)
    print(f"Sum Root to Leaf: {sum_root_to_leaf_numbers(num_root)}")

if __name__ == "__main__":
    run_examples()

File: examples/java/TreeDFSExamples.java
import java.util.*;

public class TreeDFSExamples {
    
    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }
    
    // 1. Binary Tree Inorder Traversal
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        inorder(root, result);
        return result;
    }
    
    private void inorder(TreeNode node, List<Integer> result) {
        if (node != null) {
            inorder(node.left, result);
            result.add(node.val);
            inorder(node.right, result);
        }
    }
    
    // 2. Binary Tree Preorder Traversal
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        preorder(root, result);
        return result;
    }
    
    private void preorder(TreeNode node, List<Integer> result) {
        if (node != null) {
            result.add(node.val);
            preorder(node.left, result);
            preorder(node.right, result);
        }
    }
    
    // 3. Binary Tree Postorder Traversal
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        postorder(root, result);
        return result;
    }
    
    private void postorder(TreeNode node, List<Integer> result) {
        if (node != null) {
            postorder(node.left, result);
            postorder(node.right, result);
            result.add(node.val);
        }
    }
    
    // 4. Maximum Depth of Binary Tree
    public int maxDepth(TreeNode root) {
        return root == null ? 0 : 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
    }
    
    // 5. Diameter of Binary Tree
    int diameter = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        diameter = 0;
        dfs(root);
        return diameter;
    }
    
    private int dfs(TreeNode node) {
        if (node == null) return 0;
        int left = dfs(node.left);
        int right = dfs(node.right);
        diameter = Math.max(diameter, left + right);
        return 1 + Math.max(left, right);
    }
    
    // 6. Binary Tree Maximum Path Sum
    int maxSum = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        maxSum = Integer.MIN_VALUE;
        maxGain(root);
        return maxSum;
    }
    
    private int maxGain(TreeNode node) {
        if (node == null) return 0;
        int leftGain = Math.max(maxGain(node.left), 0);
        int rightGain = Math.max(maxGain(node.right), 0);
        int pathSum = node.val + leftGain + rightGain;
        maxSum = Math.max(maxSum, pathSum);
        return node.val + Math.max(leftGain, rightGain);
    }
    
    // 7. Path Sum
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) return false;
        if (root.left == null && root.right == null) return root.val == targetSum;
        return hasPathSum(root.left, targetSum - root.val) || hasPathSum(root.right, targetSum - root.val);
    }
    
    // 8. Path Sum II
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> result = new ArrayList<>();
        dfs(root, targetSum, new ArrayList<>(), result);
        return result;
    }
    
    private void dfs(TreeNode node, int sum, List<Integer> path, List<List<Integer>> result) {
        if (node == null) return;
        path.add(node.val);
        if (node.left == null && node.right == null && sum == node.val) {
            result.add(new ArrayList<>(path));
        } else {
            dfs(node.left, sum - node.val, path, result);
            dfs(node.right, sum - node.val, path, result);
        }
        path.remove(path.size() - 1);
    }
    
    // 9. Binary Tree Paths
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> result = new ArrayList<>();
        if (root != null) dfs(root, "", result);
        return result;
    }
    
    private void dfs(TreeNode node, String path, List<String> result) {
        if (node.left == null && node.right == null) {
            result.add(path + node.val);
        } else {
            if (node.left != null) dfs(node.left, path + node.val + "->", result);
            if (node.right != null) dfs(node.right, path + node.val + "->", result);
        }
    }
    
    // 10. Path Sum III
    public int pathSumIII(TreeNode root, int targetSum) {
        if (root == null) return 0;
        return dfsPathSum(root, targetSum) + pathSumIII(root.left, targetSum) + pathSumIII(root.right, targetSum);
    }
    
    private int dfsPathSum(TreeNode node, long sum) {
        if (node == null) return 0;
        return (node.val == sum ? 1 : 0) + dfsPathSum(node.left, sum - node.val) + dfsPathSum(node.right, sum - node.val);
    }
    
    public static void main(String[] args) {
        TreeDFSExamples examples = new TreeDFSExamples();
        
        System.out.println("=== Tree DFS Examples ===");
        
        // Create test tree: [3,9,20,null,null,15,7]
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(9);
        root.right = new TreeNode(20);
        root.right.left = new TreeNode(15);
        root.right.right = new TreeNode(7);
        
        // Test 1: Binary Tree Inorder Traversal
        System.out.println("1. Inorder Traversal: " + examples.inorderTraversal(root));
        
        // Test 2: Binary Tree Preorder Traversal
        System.out.println("2. Preorder Traversal: " + examples.preorderTraversal(root));
        
        // Test 3: Binary Tree Postorder Traversal
        System.out.println("3. Postorder Traversal: " + examples.postorderTraversal(root));
        
        // Test 4: Maximum Depth of Binary Tree
        System.out.println("4. Max Depth: " + examples.maxDepth(root));
        
        // Test 5: Diameter of Binary Tree
        System.out.println("5. Diameter: " + examples.diameterOfBinaryTree(root));
        
        // Test 6: Binary Tree Maximum Path Sum
        System.out.println("6. Max Path Sum: " + examples.maxPathSum(root));
        
        // Test 7: Path Sum
        System.out.println("7. Has Path Sum (12): " + examples.hasPathSum(root, 12));
        
        // Test 8: Path Sum II
        System.out.println("8. Path Sum II (12): " + examples.pathSum(root, 12));
        
        // Test 9: Binary Tree Paths
        System.out.println("9. Binary Tree Paths: " + examples.binaryTreePaths(root));
        
        // Test 10: Path Sum III
        System.out.println("10. Path Sum III (3): " + examples.pathSumIII(root, 3));
    }
}